File: backtracking/131.py
from typing import List

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        path = []
        ans = []

        def back_tracking(start_index):
            if start_index == len(s)+1: 
                ans.append(path.copy())
                return

            begin = start_index-1
            for i in range(start_index, len(s)+1):
                if not self.is_palindrom(s[begin:i]): continue
                path.append(s[begin:i])
                back_tracking(i+1)
                path.pop()

        back_tracking(1)

        return ans


    def is_palindrom(self, s: str):
        l_ptr = 0
        r_ptr = len(s)-1
        while l_ptr < r_ptr:
            if(s[l_ptr] != s[r_ptr]): return False
            l_ptr += 1
            r_ptr -= 1
        return True


if __name__ == '__main__':
    s = Solution()
    print(s.partition("aab"))

File: backtracking/131.cc
/* 131. 分割回文串
 * 难度：中等
 */

#include <vector>
#include <string>

using namespace std;

class Solution {
private:
    vector<vector<string>> ans;
    vector<string> path;
    bool isPalindrome(string s) {
        vector<char> stack;       
        if(s.size() == 1) return true;

        for(auto c: s) {
            if(stack.size()) {
                if(stack.back() == c) stack.pop_back();
                else stack.push_back(c);
            } else {
                stack.push_back(c);
            }
        }

        if(stack.size()) return false;
        else return true;
    }

    void backtracking()
public:
    vector<vector<string>> partition(string s) {
        
    }
};
